pageserver/src/tenant.rs

```diff
@@ -1593,7 +1596,7 @@ impl Tenant {
                     .await?
             }
             None => {
-                self.bootstrap_timeline(new_timeline_id, pg_version, ctx)
+                self.bootstrap_timeline(new_timeline_id, pg_version, cancel, ctx)
```

Comment:
I am however unsure if this should have http request guided cancellation.. In the case of `timeline_create` going long (as they seem to go), then we'd have to retry and re-run these, and they'd always go long.

Might be better to only cancel with the tenant cancellation token?

Reply:
I see where you are coming from, as this is an operation that will likely be retried and if we ignore the http connection closing we give them what they want. But if the other side hangs up, the pageserver should be fully within its right to also give up on its own.

Reply:
Agreed *if* we were creating an API for whatever usage. This is however control plane with it's retry strategies. We should not start to take a end-user visible retry-loop hit because we can.

Reply:
IDK, we can not have cancellation tokens here, but ideally console has high enough timeouts that it doesn't matter.

It's not risk-free from a correctness POV though, as in you likely expect a timeline deletion to actually delete the timeline, but maybe there is some long running initdb runi in the background that finishes after your timeline deletion request finishes. But I would only call that a risk, not a full issue, because I doubt the PS API is hit in that fashion.

Only the tenant token is definitely not enough though, we also need to cancel with the global shutdown token to not have this slow down shutdowns.

Reply:
> IDK, we can not have cancellation tokens here, but ideally console has high enough timeouts that it doesn't matter.

My gut feeling is that are often requests which the control plane has to retry.

> Only the tenant token is definitely not enough though, we also need to cancel with the global shutdown token to not have this slow down shutdowns.

Surely each tenant cancellation token is a sub-token of the global one? Why is this relevant here, even if it was not, then each tenant will participate in the global shutdown just the same.

Earlier I suggested to use the tenant cancellation token. Here, with a suggestion:

```suggestion
                self.bootstrap_timeline(new_timeline_id, pg_version, &self.cancel, ctx)
```

But I cannot suggest undoing all of the changes leading up to this in github-suggestion format.

> It's not risk-free from a correctness POV though, as in you likely expect a timeline deletion to actually delete the timeline, but maybe there is some long running initdb run in the background that finishes after your timeline deletion request finishes.

There is now a test case for this case introduced in #5917 for a production issue.

Timeline creation should also hold the tenant gate open, which I am thinking it already did.

Reply:
Isn't the idea to use the TLH cancel token ?

Reply:
> Surely each tenant cancellation token is a sub-token of the global one? Why is this relevant here, even if it was not, then each tenant will participate in the global shutdown just the same.

I didn't know that.

> There is now a test case for this case introduced in https://github.com/neondatabase/neon/pull/5917 for a production issue.

Very good that we have this :+1: .

Reply:
> Isn't the idea to use the TLH cancel token ?

Here I don't think it makes sense, we could take a long time. If we were to cancel on requests, then it might be we never complete because of all of the rework. Note that we have not cancelled for the request cancel token before at all.

There was a time when we did, but that going back to that would likely cause other problems at control plane as well (ever since adding of the spawned request handlers control plane has had to make corner-case adjustments to their retry code to adapt to our new behaviour).

Reply:
27c5c4eca51fe2bef379917174a2f68085ca5f75